src/asl.ts

```diff
@@ -2129,33 +2319,152 @@ export class ASL {
         throw new SynthError(
           ErrorCodes.Unsupported_Feature,
           `Step Function does not support ${expr.op} operator`
         );
-      }
-      assertNever(expr.op);
-    } else if (isAwaitExpr(expr)) {
-      return this.eval(expr.expr);
-    } else if (isTypeOfExpr(expr)) {
-      return this.evalExpr(expr.expr, (exprOutput) => {
-        if (ASLGraph.isLiteralValue(exprOutput)) {
-          return {
-            value: typeof exprOutput.value,
-            containsJsonPath: false,
-          };
-        }
-
-        const tempHeap = this.newHeapVariable();
+      } else if (expr.op === "&&" || expr.op === "||" || expr.op === "??") {
+        return this.evalExpr(expr.left, (leftOutput) => {
+          const right = this.eval(expr.right);
+
+          if (ASLGraph.isLiteralValue(leftOutput)) {
+            // if both values are literals, return a literal.
+            // Only evaluate right as a literal if it is strictly a literal with no states.
+            if (ASLGraph.isLiteralValue(right)) {
+              return {
+                value:
+                  expr.op === "&&"
+                    ? leftOutput.value && right.value
+                    : expr.op === "||"
+                    ? leftOutput.value || right.value
+                    : leftOutput.value ?? right.value,
+                containsJsonPath: false,
+              };
+            }
 
-        return {
-          startState: "choose",
+            /**
+             * If left is a literal, evaluate the truthiness and return left or right.
+             * &&: when truthy, return right
+             * ||: when falsy, return right
+             * ??: when undefined or null, return right
+             */
+            if (expr.op === "&&") {
+              return !leftOutput.value ? leftOutput : right;
+            } else if (expr.op === "||") {
+              return !!leftOutput.value ? leftOutput : right;
+            } else {
+              return leftOutput.value !== null && leftOutput.value !== undefined
+                ? leftOutput
+                : right;
+            }
+          }
+
+          /**
+           * If the right is strictly a condition (no states), return a condition.
+           */
+          if (ASLGraph.isConditionOutput(right)) {
+            if (ASLGraph.isConditionOutput(leftOutput)) {
+              return {
+                condition:
+                  expr.op === "&&"
+                    ? ASL.and(leftOutput.condition, right.condition)
+                    : expr.op === "||"
+                    ? ASL.or(leftOutput.condition, right.condition)
+                    : // ??: a boolean cannot be undefined or null, return the left side.
+                      leftOutput.condition,
+              };
+            }
+          }
+
+          const condition =
+            /**
+             * Left is a condition
+             * ??: checks for null or undefined, a condition will always be true or false
```

Comment:
Why is this true?

Reply:
`??` looks for null or undefined. `true` and `false` are always defined and non-null. Refactored a bit.